File: Nearest Smaller Tower[ 24-April-2023]/Solution.java
//  >> JAVA CODE <<


int [] nearestSmallestTower(int [] arr){	  
        int n = arr.length;
        int[] left = new int[n];
        int[] right = new int[n];
        Stack<Integer> stack = new Stack<Integer>();
        
        // fill left array
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && arr[stack.peek()] >= arr[i]) {
                stack.pop();
            }
            if (!stack.isEmpty()) {
                left[i] = stack.peek();
            } else {
                left[i] = -1;
            }
            stack.push(i);
        }
        
        // clear the stack and fill right array
        stack.clear();
        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && arr[stack.peek()] >= arr[i]) {
                stack.pop();
            }
            if (!stack.isEmpty()) {
                right[i] = stack.peek();
            } else {
                right[i] = n;
            }
            stack.push(i);
        }
        
        int[] ans = new int[n];
        for (int i = 0; i < n; i++) {
            if (left[i] == -1 && right[i] == n) {
                ans[i] = -1;
            } else if (left[i] == -1) {
                ans[i] = right[i];
            } else if (right[i] == n) {
                ans[i] = left[i];
            } else if (i - left[i] == right[i] - i) {
                if (arr[left[i]] < arr[right[i]]) {
                    ans[i] = left[i];
                } else if (arr[left[i]] == arr[right[i]]) {
                    ans[i] = Math.min(left[i], right[i]);
                } else {
                    ans[i] = right[i];
                }
            } else {
                if (i - left[i] < right[i] - i) {
                    ans[i] = left[i];
                } else {
                    ans[i] = right[i];
                }
            }
        }
        return ans;
    }

File: Nearest Smaller Tower[ 24-April-2023]/solution.py
#     >> PYTHON CODE <<


def nearestSmallestTower(self,arr):
        n=len(arr)
        
        stack=[]
        left=[-1]*n
        right=[n]*n
        
        stack.append(0)
        
        for i in range(1,n):
            while(len(stack)!=0 and arr[stack[-1]]>=arr[i]):
                stack.pop()
            
            if(len(stack)!=0): left[i]=stack[-1]
            else: left[i]=-1
            
            stack.append(i)
        

        stack=[]
        stack.append(n-1)
        for i in range(n-2,-1,-1):
            while(len(stack)!=0 and arr[stack[-1]]>=arr[i]):
                stack.pop()
            
            if(len(stack)!=0): right[i]=stack[-1]
            else: right[i]=n
        
            stack.append(i)
        
        #print(left, right)
        ans=[]
        
        for i in range(0,n):
            if(left[i]==-1 and right[i]==n):
                ans.append(-1)
            elif left[i]==-1:
                ans.append(right[i])
            elif right[i]==n:
                ans.append(left[i])
            elif i-left[i]==right[i]-i:
                if(arr[left[i]]<arr[right[i]]):
                    ans.append(left[i])
                elif arr[left[i]]==arr[right[i]]:
                    ans.append(min(left[i],right[i]))
                else:
                    ans.append(right[i])
            else:
                if(i-left[i]<right[i]-i):
                    ans.append(left[i])
                else:
                    ans.append(right[i])
            
        return ans